[2201.11.0] Incorporate review comments (#9655)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.11.0/RELEASE_NOTE.md

## Ballerina library updates

### New features

#### `http` package

> - Added relaxed binding support for service and client data binding

Reviewer: ```suggestion
- Added relaxed binding support for service and client data binding.
```

> - Added relaxed binding support for service and client data binding

Reviewer 2: ```suggestion
- Added relaxed binding support for service and client data binding. This provides the flexibility to handle `nil` values as optional and absent fields as `nilable` types in the data projection.
```

> - Added relaxed binding support for service and client data binding.

Reviewer 2: ```suggestion
- Added relaxed binding support for service and client data binding. This provides the flexibility to assign `nil` values to optional fields and absent values to nilable fields in the data projection.
  ```ballerina
  // Enable relaxed data binding in the client side
  http:Client httpClient = check new("http://localhost:9090", laxDataBinding = true);

  // Enable relaxed data binding in the server side
  @http:ServiceConfig {laxDataBinding: true}
  service /api on new http:Listener(9090) {
  }
  ```
```

Reviewer: ````suggestion
- Added relaxed binding support for service and client data binding. This provides the flexibility to bind nil values to optional fields and absent values to nilable fields.

  ```ballerina
  // Enable relaxed data binding on the client side.
  http:Client httpClient = check new("http://localhost:9090", laxDataBinding = true);
  
  // Enable relaxed data binding on the server side.
  @http:ServiceConfig {laxDataBinding: true}
  service /api on new http:Listener(9090) {
  }
  ```
````

### Improvements

#### `http` package

- Added `anydata` support for `setPayload` methods in the request and response objects.
- Improved the `@http:Query` annotation to overwrite query parameter names in clients.
- Improved the `@http:Query` annotation to overwrite query parameter names in services.
- Added header name mapping support in record fields.

> - Migrated client and service data binding to use the `toJson` and `parserAsType` functions from the  `ballerina/data.jsondata` module instead of the `fromJsonWithType` function from the `ballerina.lang.value` module.

Reviewer 2: ```suggestion
- Migrated client and service data binding to use the `toJson` and `parserAsType` functions from the  `ballerina/data.jsondata` module instead of the `fromJsonWithType` function from the `ballerina.lang.value` module. This allows integrating the `jsondata:Name` annotation in the JSON to Ballerina record conversion.
```

> - Migrated client and service data binding to use the `toJson` and `parserAsType` functions from the  `ballerina/data.jsondata` module instead of the `fromJsonWithType` function from the `ballerina.lang.value` module, to support flexible data binding.

Reviewer: Do we really need this though? @TharmiganK Already covered under new features?

Reviewer: What's the improvement then?

Reviewer 2: This allows us to utilize the name annotation to specify different name for the fields rather than the one defined in the type. This is useful when creating record types from OpenAPI schemas where we use Ballerina friendly names for record fields without breaking the data-binding.

Additionally, this migration is a requirement for the relaxed data-binding

Reviewer: I think we should mention those. Just the switch isn't an improvement, right? Since these functions are for two different usecases.

Reviewer 2: Mentioned via [d67c961](https://github.com/ballerina-platform/ballerina-dev-website/pull/9655/commits/d67c961656993618c440e9a572fdece777e74b0f)

> - Migrated client and service data binding to use the `toJson` and `parserAsType` functions from the  `ballerina/data.jsondata` module instead of the `fromJsonWithType` function from the `ballerina.lang.value` module, to support flexible data binding.

Reviewer 2: ```suggestion
- Migrated client and service data binding to use the `toJson` and `parserAsType` functions from the  `ballerina/data.jsondata` module instead of the `fromJsonWithType` function from the `ballerina.lang.value` module.
```

## Ballerina library updates

### New features

#### `http` package

- Added relaxed binding support for service and client data binding. This provides the flexibility to bind nil values to optional fields and absent values to nilable fields.

  ```ballerina
  // Enable relaxed data binding on the client side.
  http:Client httpClient = check new("http://localhost:9090", laxDataBinding = true);
  
  // Enable relaxed data binding on the server side.
  @http:ServiceConfig {laxDataBinding: true}
  service /api on new http:Listener(9090) {
  }
  ```

### Improvements

#### `http` package

- Added `anydata` support for `setPayload` methods in the request and response objects.
- Improved the `@http:Query` annotation to overwrite query parameter names in clients.
- Improved the `@http:Query` annotation to overwrite query parameter names in services.
- Added header name mapping support in record fields.

> - Migrated client and service data binding to use the `toJson` and `parserAsType` functions from the  `ballerina/data.jsondata` module instead of the `fromJsonWithType` function from the `ballerina.lang.value` module. This allows integrating the `jsondata:Name` annotation in the JSON to Ballerina record conversion.

Reviewer 2: ```suggestion
- Migrated client and service data binding to use the `toJson` and `parserAsType` functions from the  `ballerina/data.jsondata` module instead of the `fromJsonWithType` function from the `ballerina.lang.value` module. This change improves how JSON data is converted to Ballerina records and vice versa, by allowing field name overwriting using the `jsondata:Name` annotation.
```

Reviewer 2: @MaryamZi Is this ok?

Reviewer: ```suggestion
- Migrated client and service data binding to use the `toJson` and `parserAsType` functions from the  `ballerina/data.jsondata` module instead of the `fromJsonWithType` function from the `ballerina.lang.value` module. This change improves how JSON data is converted to Ballerina records and vice versa, by allowing field names to be overridden using the `jsondata:Name` annotation.
```
